### PR TITLE
Rewrite/strapi

### PR DESCRIPTION
**Pull request by Aurelsicoko**
The core of Strapi re-designed from scratch! Some improvements have been made:

- Remove dependencies to ensure the core is light as possible.
- Load hook/middleware asynchronously to enable fast boot.
- Simplify/reduce the codebase to facilitate new contributions.
- Update configurations architecture to fit better with the API needs.
  - New file `custom.json` to have a dedicated place to put the custom configurations.
  - Allow hook configurations by environment.
  - New file `response.json`, `request.json`.
  - `i18n.json` renamed `language.json`
  - `general.json` renamed `application.json`
  - `hooks.json` renamed `hook.json` and splitted with `middleware.json` (new file)
- Allow custom middlewares/hooks to be loaded directly from the app.
- Harmonize database settings between connectors.
- Harmonize configurations to keep consistency and logic into them.

### CODE REVIEW

File: packages/strapi-ejs/lib/index.js [hunk 1]
```diff
@@ -37,10 +38,10 @@ module.exports = function (strapi) {
     initialize: cb => {
       // Force cache mode in production
       if (strapi.config.environment === 'production') {
-        strapi.config.hooks.ejs.cache = true;
+        strapi.config.hook.ejs.cache = true;
```
Comment (Aurelsicoko) [inline packages/strapi-ejs/lib/index.js]: `strapi.config.hook.settings.ejs.cache`

File: packages/strapi-generate-plugin/files/admin/src/containers/NotFoundPage/index.js~74aef739ab1a971f3ff26f018b672577305d21a4 [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/**
```
Comment (Aurelsicoko) [inline packages/strapi-generate-plugin/files/admin/src/containers/NotFoundPage/index.js~74aef739ab1a971f3ff26f018b672577305d21a4]: Delete this file

File: packages/strapi-generate-plugin/files/admin/src/containers/NotFoundPage/index.js~HEAD [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/**
```
Comment (Aurelsicoko) [inline packages/strapi-generate-plugin/files/admin/src/containers/NotFoundPage/index.js~HEAD]: Delete this file

File: packages/strapi-mongoose/lib/index.js [hunk 4]
```diff
@@ -46,7 +46,7 @@ module.exports = function (strapi) {
       }
 
       _.forEach(_.pickBy(strapi.config.connections, {connector: 'strapi-mongoose'}), (connection, connectionName) => {
-        const {host, port, username, password, database} = _.defaults(connection.settings, strapi.hooks.mongoose.defaults);
+        const {host, port, username, password, database} = _.defaults(connection.settings, strapi.config.hook.settings);
```
Comment (Aurelsicoko) [inline packages/strapi-mongoose/lib/index.js]: `strapi.config.hook.settings.mongoose`?

File: packages/strapi-redis/lib/index.js [hunk 5]
```diff
@@ -39,17 +40,27 @@ module.exports = function(strapi) {
...
+
       const done = _.after(_.size(connections), () => {
         cb();
       });
 
       // For each connection in the config register a new Knex connection.
       _.forEach(connections, (connection, name) => {
         // Apply defaults
-        _.defaults(connection.settings, strapi.hooks.redis.defaults);
+        _.defaults(connection.settings, strapi.hook.redis.defaults);
```
Comment (Aurelsicoko) [inline packages/strapi-redis/lib/index.js]: `strapi.config.hook.settings.redis` should be better to keep consistency between hooks.

File: packages/strapi/lib/Strapi.js [hunk 6]
```diff
@@ -1,104 +1,205 @@
...
-  // Method to expose instance globals
-  exposeGlobals(cb) {
-    require('./private/exposeGlobals').apply(this, [cb]);
+  async load() {
+    // Create AST.
+    await Promise.all([
+      nestedConfigurations.call(this),
+      apis.call(this),
+      plugins.call(this),
+      // admin.call(this),
```
Comment (Aurelsicoko) [inline packages/strapi/lib/Strapi.js]: Remove the `admin` because we don't need it anymore.

**Comment by pierreburgy** [comment]
`packages/strapi-generate-new/files/config/environments/test/hook.json` should not be empty.

**Comment by pierreburgy** [comment]
I am not able to generate a new API:
```
➜  my-app strapi generate:api user
error: Error: ENOENT: no such file or directory, open '/Users/pierreburgy/Sites/sandbox/my-app/config/environments/development/databases.json'
    at Object.fs.openSync (fs.js:651:18)
    at Object.fs.readFileSync (fs.js:553:33)
    at Object.module.exports [as before] (/Users/pierreburgy/Repositories/strapi/packages/strapi-generate-api/lib/before.js:113:38)
    at generate (/Users/pierreburgy/Repositories/strapi/packages/strapi-generate/lib/generate.js:45:13)
    at module.exports (/Users/pierreburgy/Repositories/strapi/packages/strapi-generate/lib/index.js:61:3)
    at module.exports (/Users/pierreburgy/Repositories/strapi/packages/strapi/bin/strapi-generate.js:50:10)
    at Command.program.command.option.option.option.description.action (/Users/pierreburgy/Repositories/strapi/packages/strapi/bin/strapi.js:70:33)
    at Command.listener (/Users/pierreburgy/Repositories/strapi/packages/strapi-utils/node_modules/commander/index.js:300:8)
    at emitTwo (events.js:125:13)
    at Command.emit (events.js:213:7)        `invalid` case triggered, but no handler was implemented.
```

File: packages/strapi/lib/core/configurations.js [hunk 7]
```diff
@@ -0,0 +1,299 @@
+'use strict';
+
+// Dependencies.
+const path = require('path');
+const glob = require('glob');
+const utils = require('../utils');
+const { difference, merge, setWith, get, set, upperFirst, isString, isEmpty, isObject, orderBy, isBoolean, pullAll, defaults, mapValues, isPlainObject, reduce } = require('lodash');
```
Comment (Aurelsicoko) [inline packages/strapi/lib/core/configurations.js]: Good catch!

File: packages/strapi/lib/core/configurations.js [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+      );
+
+      resolve();
+    })
+  ]);
+};
+
+module.exports.app = async function() {
+    // Retrieve Strapi version.
+    this.config.info.strapi = (get(this.config, 'info.dependencies.strapi') || '').replace(/(\^|~)/g, ''),
```
Comment (Aurelsicoko) [inline packages/strapi/lib/core/configurations.js]: No f*****g idea. It fixed!

**Comment by pierreburgy** [comment]
`User is not defined` after new API creation.

Steps to reproduce:
 - Create a new project named `my-app`: `strapi new my-app`.
 - Generate a new api: `strapi generate:api user`.
 - Start the server.
 - Visit http://localhost:1337/user.

Error stack trace:
```
info: Server started in /Users/pierreburgy/Sites/sandbox/my-app
info: Your server is running at http://localhost:1337
debug: Time: Mon Jul 31 2017 16:39:50 GMT+0200 (CEST)
debug: Launched in: 262 ms
debug: Environment: development
debug: Process PID: 20243
debug: Version: 3.0.0-alpha.4.8 (node v8.1.4)
info: To shut down your server, press <CTRL> + C at any time
error:  ReferenceError: User is not defined
    at Object.fetchAll (/Users/pierreburgy/Sites/sandbox/my-app/api/user/services/User.js:21:12)
    at find (/Users/pierreburgy/Sites/sandbox/my-app/api/user/controllers/User.js:18:45)
    at dispatch (/Users/pierreburgy/Repositories/strapi/packages/strapi/node_modules/koa-router/node_modules/koa-compose/index.js:44:32)
    at next (/Users/pierreburgy/Repositories/strapi/packages/strapi/node_modules/koa-router/node_modules/koa-compose/index.js:45:18)
    at dispatch (/Users/pierreburgy/Repositories/strapi/packages/strapi/node_modules/koa-compose/index.js:42:32)
    at next (/Users/pierreburgy/Repositories/strapi/packages/strapi/node_modules/koa-compose/index.js:43:18)
    at module.exports (/Users/pierreburgy/Repositories/strapi/packages/strapi/lib/middlewares/responses/policy.js:41:11)
    at dispatch (/Users/pierreburgy/Repositories/strapi/packages/strapi/node_modules/koa-compose/index.js:42:32)
    at next (/Users/pierreburgy/Repositories/strapi/packages/strapi/node_modules/koa-compose/index.js:43:18)
    at /Users/pierreburgy/Repositories/strapi/packages/strapi/lib/middlewares/router/utils/routerChecker.js:29:11
debug: GET /user (6ms)
```

**Comment by pierreburgy** [comment]
@Aurelsicoko I solved the issue `User is not defined` by running `npm link strapi-mongoose` in my project. It could be a good idea to add a warning message saying that a hook is not ready.

**Comment by Aurelsicoko** [comment]
It's done! The server starts though but a warning message is displayed to inform the user that he should run `npm install strapi-mongoose`.

**Comment by pierreburgy** [comment]
LGTM.

**Comment by lauriejim** [comment]
LGTM.

**Pull request merged**